### PR TITLE
Make OCSAgents robust against short duration crossbar dropouts

### LINKED ISSUE #144: Restarting the crossbar server requires restarting all agents
Agents should handle a quick restart of the crossbar server. While crossbar restarts are rare, they need to be handled properly by connected Agents.

### PR DESCRIPTION
**Pull request by BrianJKoopman**
## Description

There's still some work to be done on the integration testing and how that's organized, but I wanted to open the PR to mostly discuss the changes to the core OCS parts like `ocs_agent.py` and `ocs_feed.py`. There are many things going on here, here's a quick summary of changes.

### Agents
* Add log level setting via environment variables to Aggregator and Fake Data Agents (and accompanying documentation)
* Document session.data for `start_aggregate` process in Aggregator
* Add frame length argument for setting the frame length passed as an agg_param by the Fake Data Agent

### `ocs_agent.py`
* Perform startup operations only on first time joining the crossbar server realm
* Track when the Agent has joined and left the crossbar realm (effectively tracking whether the crossbar server is online or not)
* Stop running sessions normally if shutdown is considered normal (`wamp.close.normal`)
  * This sort of implies not shutting down the sessions if shutdown is not normal, not normal being the transport is lost (i.e. crossbar connection down)
* Clear the subscribed_topics set onLeave to force a re-subscription next time we join the crossbar realm.
* On disconnect from the crossbar server, check for reconnection for 10 seconds. Successful reconnection is logged and no other action taken. onJoin, etc. triggered again as expected
* On disconnect from the crossbar server if the reconnection 10 second timer timesout, stop running processes and stop the reactor, which should cleanly exit the Agent (if it doesn't, likely something in the Agent is stuck in a loop or similar)
* Add try/excepts for `TransportLost` exception on publishing the status to the crossbar server. This now just logs errors that indicate the crossbar server is likely down.

### `ocs_feed.py`
* Add similar `TransportLost` exception handling to `ocs_feed`

### Testing
* Introduce integration testing which relies on docker and docker-compose, using the pytest-docker-compose plugin
  * This still needs some work to properly setup and document more generally/for the CI environment

## Motivation and Context

Agents can't currently reconnect to a disconnected crossbar server. We should be able to tolerate short term reboots, but want to raise some flags if the crossbar server goes down longer than that.

closes #144 

## How Has This Been Tested?

There's been a mix of testing manually and testing with the included integration tests. Testing manually is generally done by bringing up the docker container set defined in the integration testing `docker-compose.yaml` and then either restarting or stopping the crossbar container and observing the resulting interaction with the Agents.

The ability for data to make it through a reboot has been observed to be somewhat variable. This requires more investigation and testing. I would expect data published during a reboot would be lost. Some sort of buffering mechanism, if needed, should be addressed in a future PR.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

**Comment by BrianJKoopman** [comment]
Feedback from the telecon on Wednesday has been incorporated. Please take a look again now if interested.

The integration testing is generalized enough and organized so that it can be run locally. I'm going to leave setting up the CI pipeline for it to a later PR.

**Review by jlashner — APPROVED** [review]
Looks good, and I'm excited about integration tests. Are you planning on using it to add agent operation coverage tests? Anyway I'd say go ahead and merge.

File: ocs/ocs_agent.py [hunk 1]
```diff
@@ -125,6 +129,21 @@ def __init__(self, config, site_args, address=None, class_name=None):
...
+            if self.sessions[session] is not None:
+                self.log.info("Stopping session {sess}", sess=session)
+                self.log.debug("session details: {sess}",
+                               sess=self.sessions[session].encoded())
+                if session in self.tasks:
+                    yield self.abort(session)
+                elif session in self.processes:
+                    yield self.stop(session)
+        # Give a second for processes to stop cleanly
+        yield dsleep(1)
```
[the remark this answers is not in the document]
Comment (BrianJKoopman) [reply ocs/ocs_agent.py:145]: Gotcha, extended to 3 seconds, think that'll be sufficient for the cases you listed?

**Comment by BrianJKoopman** [comment]
> Looks good, and I'm excited about integration tests. Are you planning on using it to add agent operation coverage tests?

I would love it if we could get coverage information from this, but it's actually pretty tricky with separate components running in different containers. I don't even really know if it's possible with this configuration, since pytest isn't really seeing what's going on within the various containers.

**Pull request merged**